### ISSUE TITLE
Cluster Tag Feature

### ISSUE THREAD

**Issue by cavemanloverboy** [coral-xyz/anchor#2397]
Example Problem Scenario: 

1) An anchor contract is deployed to the same address on mainnet and testnet.
2) A user wants to use a program on testnet.
3) A malicious RPC provides the latest blockhash to the mainnet network.
4) A user signs the transaction, changing state on mainnet instead of testnet.

Solution:

Add a `"mainnet"` feature flag to anchor programs which which appends `"mainnet"` to the seed of the discriminator of every instruction, i.e. something like
```
let mut pre_image = discriminator_preimage.as_bytes();
#[cfg(feature = "mainnet")]
pre_image.push_str("mainnet");
discriminator.copy_from_slice(
            &anchor_syn::hash::hash(pre_image).to_bytes()[..8],
        );
```
This creates distinct instruction discriminators for mainnet vs testnet, and allow wallets to provide strict guarantees about which cluster a user is signing for when interacting with a known anchor program.

**Comment by Henry-E** [comment]
Nice idea but it's a pretty big breaking change for anyone who updates their deployed program to a newer anchor version. Maybe if we could make this change whenever we introduce some wider changes to how discriminators are working. Like there's been talk of dropping down to 4 byte discriminators and/or adding support for version numbers in those other 4 bytes.

Will try to remember this.

**Comment by jordansexton** [comment]
I thought a bit more about the problem outlined @ https://github.com/solana-labs/wallet-standard/issues/9

A possible way for wallets to defend against this is to ignore the blockhash and check if accounts referenced by the transaction (for example, the fee payer, program ids, etc) exist on Solana mainnet.

This doesn't scale well to other SVM chains, because a wallet won't know to check them all. I think dapps adding custom RPCs to wallets is very problematic.

I imagine what's likely needed is a registry of RPCs to chain IDs. Dapps should be able to request to send transactions on a given chain ID, but probably not provide the RPC URL for a known (mainnet) chain ID.

**Comment by jordansexton** [comment]
With regard to this proposal, and the idea of checking accounts, I think in the long term both are likely the wrong places to defend against chain replay attacks. [EIP-155](https://eips.ethereum.org/EIPS/eip-155) solves this by encoding the chain ID in the wire transaction format. I have suspected for a while that we might need to do this on Solana.

**Comment by cavemanloverboy** [comment]
> I thought a bit more about the problem outlined @ [solana-labs/wallet-standard#9](https://github.com/solana-labs/wallet-standard/issues/9)
> 
> A possible way for wallets to defend against this is to ignore the blockhash and check if accounts referenced by the transaction (for example, the fee payer, program ids, etc) exist on Solana mainnet.
> 
> This doesn't scale well to other SVM chains, because a wallet won't know to check them all. I think dapps adding custom RPCs to wallets is very problematic.
> 
> I imagine what's likely needed is a registry of RPCs to chain IDs. Dapps should be able to request to send transactions on a given chain ID, but probably not provide the RPC URL for a known (mainnet) chain ID.

I don't see how this solves the problem since the wallet will probably use the same malicious RPC which may return a fake account. You need something either at the contract or L1 level.

**Comment by jordansexton** [comment]
Further discussion on this @ https://discord.com/channels/428295358100013066/885976714646290473/1073664382011191396

**Comment by jordansexton** [comment]
> I don't see how this solves the problem since the wallet will probably use the same malicious RPC which may return a fake account.

No, the wallet should use trusted RPCs for this. Look at the tx you get, and ignore what the untrusted RPC says about it first. Check the programs referenced on it against your trusted RPC. If they are there, don't use the untrusted RPC at all.

**Comment by cavemanloverboy** [comment]
> > I don't see how this solves the problem since the wallet will probably use the same malicious RPC which may return a fake account.
> 
> No, the wallet should use trusted RPCs for this. Look at the tx you get, and ignore what the untrusted RPC says about it first. Check the programs referenced on it against your trusted RPC. If they are there, don't use the untrusted RPC at all.

1) Why not just use trusted RPCs to begin with?
2) Why not push for a trustless system?

**Comment by jordansexton** [comment]
> 1. Why not just use trusted RPCs to begin with?

Wallets already do. But re: the Discord discussion about this, it should be possible for wallets to add instructions/use bundles to check that it's running on the expected cluster rather than requiring individual programs to do something.

> 2. Why not push for a trustless system?

If you can't trust the RPC, light clients are probably the answer, right?

**Comment by cavemanloverboy** [comment]
> > 1. Why not just use trusted RPCs to begin with?
> 
> Wallets already do. But re: the Discord discussion about this, it should be possible for wallets to add instructions/use bundles to check that it's running on the expected cluster rather than requiring individual programs to do something.
> 
> > 2. Why not push for a trustless system?
> 
> If you can't trust the RPC, light clients are probably the answer, right?

1) I'll reread the discussion, didn't read that far down.

2) imo, for this problem at least, light clients are unnecessary compute vs just using a static pre-computed encoding.
